python/tank/platform/software_launcher.py

```diff
@@ -205,72 +235,170 @@ def display_name(self):
+
+        Example::
+
+            ["15.5.324", "16.0.1.322"]
+        """
+        return self._versions
+
+    ##########################################################################################
+    # abstract methods
 
     def prepare_launch(self, exec_path, args, file_to_open=None):
         """
         Prepares the given software for launch
 
         :param str exec_path: Path to DCC executable to launch
         :param str args: Command line arguments as strings
-        :param str file_to_open: (optional) Full path name of a file to open on launch
+        :param str file_to_open: (optional) Full path name of a file to open on
+            launch
         :returns: :class:`LaunchInformation` instance
         """
         raise NotImplementedError
 
+    def _glob_and_match(self, match_template, tokens_expressions):
+        """
+        This is a helper method that can be invoked in an implementation of :meth:`scan_software`.
+
+        The ``match_template`` argument provides a template to use both for globbing files and then pattern
+        matching them using regular expressions provided by the ``tokens_expressions`` dictionary.
+
+        The method will first substitute every token surrounded by `{}` from the template with a ``*``
+        for globbing files. It will then replace the tokens in the template with the regular expressions
+        that were provided.
+
+        In the following example::
+
+            self._glob_and_match(
+                "//network/softwares/Nuke{full_version}/Nuke{major_minor_version}.exe",
+                {
+                    "full_version": r"([\d.v]+)",
+                    "major_minor_version": r"([\d.]+)"
+                }
+            )
+
+        this would first look for every files matching C:/Program Files/Nuke*/Nuke*.exe and then
+        run the regular expression C:/Program Files/Nuke(?P<version>[\d.v]+)/(?P<major_minor_version>[\d.]+)
+        and would return any files matching that pattern as well as any extracted tokens.
+
+        .. note::
+            The ``match_template`` argument must always use ``/`` for path separators. This allows to disambiguate
+            the meaning of `\\`` on Windows when matching executables with regular expressions.
+
+        :param str match_template: Template that will be used both for globbing and performing a regular expression.
+
+        :param dict tokens_expressions: Dictionary of regular expressions that can be substituted
+            in the template. The key should be the name of the token to substitute.
+
+        :returns: A list of tuples containing the path, the groups and the group dictionary matches
+            from the regular expression. For example, if Nuke 10.0v1 had been installed on the network share
+            provided above, the return value would have been::
+                [("//network/softwares/Nuke10.0v1/Nuke10.1.exe",
+                  ("10.0v1", "10.0"),
```

Comment:
Feels like we should return something for every token. I wonder if we should imply the parenthesis?